FQDN IP updates wait on endpoint regenerations

raybejjani opened the pull request:
This PR supersedes https://github.com/cilium/cilium/pull/9185 (assuming it works).
I still need to test this a little more but the general structure is there. An extension would be to add an option to wait on the endpoint that made the request, all endpoints, or no endpoints. I'm not sure that it's worth it, however, since any reasonable timeout would trigger in the cases where waiting on all endpoints is too slow (which also means the single-endpoint case is also slow).

I'm also a little worried that this PR adds a dependency of a type that @ianvernon is trying to remove. In particular, we now rely on a `BumpRevision` call for IP updates to selectors, and we may have a future where such updates don't require the bump or regeneration.

When new IPs are added to the DNS cache we also update FQDN
Selectors. These, in turn, may trigger policy regenerations. The most
ideal behaviour is that we have plumbed all needed policy before we
return the DNS response to the pod.
The DNS proxy now notifies about DNS responses before it forwards the
response. The daemon handler waits for regeneration, with a timeout.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9208)
<!-- Reviewable:end -->

raybejjani committed:
policy: Expose map-update WaitGroup in FQDN update callchains

A FQDN IP update will trigger map updates. Callers, such as the DNS
proxy handlers, may wish to wait on this update to complete. This new
WaitGroup needs to be exposed to callers.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

raybejjani committed:
FQDN: Wait on policy map update when adding new IPs

When new IPs are added to the DNS cache we also update FQDN
Selectors. These, in turn, may require datapath updates. The most
ideal behaviour is that we have plumbed all needed policy before we
return the DNS response to the pod.
The DNS proxy now notifies about DNS responses before it forwards the
response and the daemon handler waits for regeneration, with a timeout.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

ianvernon reviewed (approved):
LGTM

ianvernon commented on daemon/fqdn.go:
```diff
@@ -121,9 +121,9 @@ func (d *Daemon) updateSelectorCacheFQDNs(selectors map[policyApi.FQDNSelector][
+		return 0
 	}
-	d.TriggerPolicyUpdates(false, "updated identities for FQDNs")
+	return d.TriggerPolicyUpdates(false, "updated identities for FQDNs")
```
For this PR to work this needs to be `true`. This has big performance implications, though, as this means we will be regenerating all endpoints quite frequently.

ianvernon replied:
We actually have logic which updates the in-memory policymap state for Endpoints. See `ConsumeMapChanges` - in `pkg/policy`.

I think that we shouldn't have to call `TriggerPolicyUpdates` here at all. Instead, we should have a separate type of event which updates the policy maps for all endpoints on the node. This could be an `eventqueue.Event` that we `Enqueue` onto the `EventQueue` for the endpoint. Given that there is no actual change to the policy "rules" with these FQDN updates, we wouldn't need to bump the endpoint's policy revision.

I'm recommending using the EventQueue construct here out of caution, as that guarantees that any regenerations which are in line for the endpoint are taken care of first as well. I'm wondering whether we can simply asynchronously update the policymap state for the endpoint and return a channel when all updates for Endpoints are done.

raybejjani replied:
I'm fine with anything, really, as long as we can block appropriately. I guess all FQDN updates of this sort would go through the newer policymap update pipeline so we don't need to mix that and another wait for when a new policy comes in (that wouldn't be a real-time thing anyway).

ianvernon reviewed (changes requested):
Revoking approval due to the need to change parameters for `TriggerPolicyUpdates`.

raybejjani committed:
endpoint: Expose Endpoint.ApplyPolicyMapChanges

There are circumstances where we need to update an endpoint's policy map
without any other actions, such as when we receive new IPs from the DNS
proxy. This assumes that no further policy updates are need to the
datapath. This is true when adding new IPs (i.e. identities) to an
already present endpoint policy, and can be done asynchronously.

Note: This assumes that PolicyMap updates have already been prepared via
`AccumulateMapChanges` further down the call stack. In the FQDN case
this occurs while preparing the selector updates.

Signed-off-by: Ray Bejjani <ray@isovalent.com>

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -886,6 +886,15 @@ func (e *Endpoint) addPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntry,
+// that have accumulated for the PolicyMap via various outside events (e.g.,
+// identities added / deleted).
+func (e *Endpoint) ApplyPolicyMapChanges() error {
+	e.mutex.Lock()
```
As discussed offline we should make this a `lockAlive` - not sure if we will want to log the error in that case in the caller; perhaps we can switch on the error type and only log if the error is not a liveness error.

raybejjani replied:
`UpdatePolicyMaps` warns on error. Did you mean that we would switch in there or in `ApplyPolicyMapChanges`?

ianvernon replied:
I'm ambivalent - I guess it's better to return an error and let the caller do what if wants with it?

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -886,6 +886,15 @@ func (e *Endpoint) addPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntry,
 	return true
 }
 
+// ApplyPolicyMapChanges updates the Endpoint's PolicyMap with the changes
```
Please sign off on the commit as well.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -886,6 +886,17 @@ func (e *Endpoint) addPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntry,
+	if err := e.lockAlive(); err != nil {
+		return err
+	}
+	defer e.mutex.Unlock()
```
`defer e.unlock()`?

aanm commented on daemon/fqdn.go:
```diff
@@ -121,9 +122,10 @@ func (d *Daemon) updateSelectorCacheFQDNs(selectors map[policyApi.FQDNSelector][
+		return &sync.WaitGroup{}
 	}
-	d.TriggerPolicyUpdates(false, "updated identities for FQDNs")
+	wg = d.endpointManager.UpdatePolicyMaps()
```
why do you return it directly in 125 but not here?

aanm commented on daemon/fqdn.go:
```diff
@@ -475,14 +475,36 @@ func (d *Daemon) notifyOnDNSMsg(lookupTime time.Time, ep *endpoint.Endpoint, epI
+		updateComplete := make(chan struct{})
+		go func(wg *sync.WaitGroup, done chan struct{}) {
+			wg.Wait()
+			updateComplete <- struct{}{}
```
you can simply close the channel

aanm reviewed (commented):
I think all of my request changes are small nits. not blocking on my side

jrajahalme reviewed (approved):
Bit nervous about applying the collected incremental changes out-of-band (w.r.t. regeneration), but this is the right move, and look OK :-)
